Insertion point bar: hide onBlur and onMouseLeave

ramonjd opened the pull request:
Possibly, in some way, resolves https://github.com/WordPress/gutenberg/issues/35536
Also resolves https://github.com/WordPress/gutenberg/issues/36882

## Description

This PR hides the inserter point via the `hideInsertionPoint` action when the horizontal inserter bar loses focus, or when the mouse leaves.

`hideInsertionPoint` will not fire when the block inserter panel is open.

![Kapture 2021-11-25 at 11 49 00](https://user-images.githubusercontent.com/6458278/143332274-9b2fe4eb-6bfc-4d6e-a479-668632a91cb1.gif)

We might not have covered all cases documented in https://github.com/WordPress/gutenberg/issues/35536, and there is no timeout, but it might help to avoid some of the concerns with the persistent inserter bar.

Try it out today!

## How has this been tested?

Take a look at the various scenarios over at https://github.com/WordPress/gutenberg/issues/35536

With the right mouse artistry you might find that the bar persists, but, at least in my testing, it doesn't persist as often with these changes.


## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ramonjd committed:
Hiding insertion point when a user focusses outside the insertion popover, and also the mouse leaves the horizontal line.
Only hiding when the block inserter panel is not open

jasmussen commented:
My goodness, even as is, this feels like a VAST and profound improvement over what's in trunk. As a reminder, this is how easy it is to stick the sibling inserter in trunk at the moment:
![trunk](https://user-images.githubusercontent.com/1204802/143417478-2c98970b-7567-4544-836e-5603722492fe.gif)

Here's this branch:

![this branch](https://user-images.githubusercontent.com/1204802/143417497-ef3f98c2-ba21-4580-b086-9e93783fd7d9.gif)

This feels like a massive forward step. I'd love for this to land as soon as possible! Thank you!

jasmussen reviewed (approved):
Gave it another spin, and in all my testing I saw no downsides to this one. 

The code looks minimal and good. I'd encourage just a quick sanity check from another dev, but honestly this feels good enough to land.

ntsekouras commented on packages/block-editor/src/components/block-tools/insertion-point.js:
```diff
@@ -294,6 +300,8 @@ function InsertionPointPopover( {
+			onFocusOutside={ maybeHideInserterPoint }
+			onMouseLeave={ maybeHideInserterPoint }
 		>
 			<motion.div
```
I've started looking at this and I'm not familiar with this code, but I was wondering if you have search for a solution from [motion](https://www.framer.com/docs/gestures/###onhoverend) maybe something like `onHoverEnd={ maybeHideInserterPoint }` in this motion.div?

ramonjd replied:
> I was wondering if you have search for a solution from motion maybe something like onHoverEnd={ maybeHideInserterPoint } in this motion.div?

Thanks for looking!

I tried it out and I don't see any difference in behaviour, so I don't see any implementation issues with it.

Was there a specific reason why we should prefer motion's eventhandler?

ramonjd replied:
> Was there a specific reason why we should prefer motion's eventhandler?

Oh, I did find one advantage, aside from the fact that `Popover` doesn't specifically define an `onMouseLeave` event: we can perform an equality check on the even target and the current ref to make sure the event is occurring on the motion div. 

I've pushed that change since I couldn't detect any noticeable divergence in behaviour.

ntsekouras reviewed (commented):
Thanks for working on this Ramon! It would be great if @ellatrix shared her thoughts here as she has worked a lot with these parts. Should this be better handled here: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-list/use-in-between-inserter.js#L15?

apeatling reviewed (commented):
Tried this and I agree is a big improvement. I'm not familiar with the code, so good to get @ellatrix's input if it makes sense in a difference place. Either way this should land once that's done.

ramonjd commented:
Thanks for testing, everyone!

>  It would be great if @ellatrix shared her thoughts here as she has worked a lot with these parts. Should this be better handled here: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-list/use-in-between-inserter.js#L15?

Yes, I'm also a newbie in this area of the code so would be interested to hear if there's a better way. I couldn't see how we could attach event to the inserter in that hook.

ramonjd committed:
Move the mouseleave event handler to be the motion onHoverEnd callback.
There seems to be no noticeable difference in behaviour, only that the Popover component doesn't specifically define an `onMouseLeave`.
Also, we can more easily perform a check against the ref and event target.

ramonjd commented:
I removed the `onFocusOutside` for now, mainly because in testing I couldn't detect any noticeable difference without it. 

There is an edge case that this PR does not address: the inserter persists when you click the plus icon and drag your mouse away from it. 

This is because there's an exisiting [onFocus callback](https://github.com/WordPress/gutenberg/blob/9ebdd3719248ae7fc03ad723fa6bee805017545b/packages/block-editor/src/components/block-tools/insertion-point.js#L190) on the motion div that sets `openRef.current = true;`, which is supposed to indicate that the block inserter panel is open.

ntsekouras reviewed (approved):
LGTM, thanks Ramon! Changes are minimal and make for a way better experience.

>There is an edge case that this PR does not address: the inserter persists when you click the plus icon and drag your mouse away from it.

I think we can address this in a follow up, as it's already in trunk. Maybe this should be handled in `useInBetweenInserter` as we are only listening to `mousemove` event and we could add logic for detecting the dragging.
